From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 16 May 2024 06:40:03 +0300
Subject: [PATCH] Implement packet item faker for injected items

Forces default item faker for injected items as to not kick the client from the server because of modded items

diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
index 573c389779c581a70d9196596b4c63d587e9ffd2..249fd0f184fede1cc90464524eb43fca508b096c 100644
--- a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
@@ -221,7 +221,11 @@ public class ItemModifiersHandlerImpl extends ItemModifiersHandler {
 	}
 
 	public static void init() {
+        ItemModifier itemFaker = modifiers.get(KiterinoItemModifierImpl.MODIFIER_KEY);
 		modifiers.clear(); // In case of reload. Not supported, but eh
+        if (itemFaker != null) {
+            modifiers.put(itemFaker.getModifierId(), itemFaker);
+        }
 		if (ItemModifiersHandler.itemModifiersHandler == null) ItemModifiersHandler.itemModifiersHandler = new ItemModifiersHandlerImpl();
 		ItemPacketContextInitializer.init();
 	}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/KiterinoItemModifierImpl.java b/src/main/java/me/sosedik/kiterino/modifier/item/KiterinoItemModifierImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..16242986d124173007c4e18dd6219e5d9e9f8061
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/KiterinoItemModifierImpl.java
@@ -0,0 +1,39 @@
+package me.sosedik.kiterino.modifier.item;
+
+import org.bukkit.NamespacedKey;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Handles visual modification for injected items
+ */
+@ApiStatus.Internal
+public final class KiterinoItemModifierImpl extends ItemModifier {
+
+    public static final NamespacedKey MODIFIER_KEY = new NamespacedKey(NamespacedKey.KITERINO, "item_faker");
+    public static final KiterinoItemModifierImpl MODIFIER_IMPL = new KiterinoItemModifierImpl();
+
+    private final Map<NamespacedKey, KiterinoItemModifier> modifiers = new HashMap<>();
+
+    private KiterinoItemModifierImpl() {
+        super(MODIFIER_KEY);
+    }
+
+    public void add(@NotNull NamespacedKey key, @NotNull KiterinoItemModifier modifier) {
+        this.modifiers.put(key, modifier);
+    }
+
+    @Override
+    public @NotNull ModificationResult modify(@NotNull ItemContextBox contextBox) {
+        KiterinoItemModifier modifier = modifiers.get(contextBox.getItem().getType().getKey());
+        if (modifier != null) {
+            modifier.modify(contextBox);
+            return ModificationResult.OK;
+        }
+        return ModificationResult.PASS;
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
index 53885600bdd389949aca90a01a039737e4dd3a0b..2a60bf487fe0a5e6cb7be88fef2d68078e9af3ce 100644
--- a/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
+++ b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
@@ -1,5 +1,12 @@
 package me.sosedik.kiterino.registry.data;
 
+import me.sosedik.kiterino.modifier.item.KiterinoItemModifier;
+import me.sosedik.kiterino.modifier.item.KiterinoItemModifierImpl;
+import net.minecraft.core.component.DataComponents;
+import org.bukkit.inventory.ItemStack;
+import net.minecraft.world.item.component.UseCooldown;
+import java.util.Optional;
+
 import net.minecraft.core.registries.BuiltInRegistries;
 
 import io.papermc.paper.datacomponent.DataComponentType;
@@ -23,6 +30,7 @@ public class KiterinoItemRegistryEntity implements ItemRegistryEntity {
     protected Map<DataComponentType, Object> components = new HashMap<>();
     protected @Nullable Float compostChance;
     protected @Nullable Item nmsItem;
+    protected @Nullable KiterinoItemModifier modifier;
 
     protected final Conversions conversions;
     public ResourceLocation itemKey;
@@ -63,7 +71,8 @@ public class KiterinoItemRegistryEntity implements ItemRegistryEntity {
     @Override
     public @NonNull Object constructItemProperties() {
         return new Item.Properties()
-                .setId(ResourceKey.create(Registries.ITEM, this.itemKey));
+                .setId(ResourceKey.create(Registries.ITEM, this.itemKey))
+                .component(DataComponents.USE_COOLDOWN, new UseCooldown(0F, Optional.of(this.itemKey)));
     }
 
     @Override
@@ -76,6 +85,11 @@ public class KiterinoItemRegistryEntity implements ItemRegistryEntity {
         return BuiltInRegistries.BLOCK.getValueOrThrow(ResourceKey.create(Registries.BLOCK, this.itemKey));
     }
 
+    @Override
+    public @Nullable KiterinoItemModifier modifier() {
+        return this.modifier;
+    }
+
     protected void parseValues(@NonNull Item item) {
         // Does it make sense to parse?
         // DataComponentMap components = item.components();
@@ -114,6 +128,33 @@ public class KiterinoItemRegistryEntity implements ItemRegistryEntity {
             return this;
         }
 
+        @Override
+        public @NonNull Builder modifier(@Nullable KiterinoItemModifier modifier) {
+            this.modifier = modifier;
+            return this;
+        }
+
+        @SuppressWarnings("unchecked")
+        private @NonNull KiterinoItemModifier buildModifier(@NonNull Item nmsItem) {
+            return box -> {
+                if (this.modifier != null) this.modifier.modify(box);
+                ItemStack item = box.getItem();
+                // TODO this is dumb
+                nmsItem.components().keySet().forEach(dataType -> {
+                    if (dataType == DataComponents.ITEM_MODEL) return;
+
+                    DataComponentType type = PaperDataComponentType.minecraftToBukkit(dataType);
+                    if (item.isDataOverridden(type)) return;
+
+                    if (type instanceof DataComponentType.Valued<?> valued) {
+                        item.setData((DataComponentType.Valued<Object>) valued, PaperDataComponentType.convertDataComponentValue(nmsItem.components(), (PaperDataComponentType.ValuedImpl) valued));
+                    } else if (type instanceof DataComponentType.NonValued nonValued) {
+                        item.setData(nonValued);
+                    }
+                });
+            };
+        }
+
         @SuppressWarnings("unchecked")
         @Override
         public @NonNull Item build() {
@@ -130,6 +171,7 @@ public class KiterinoItemRegistryEntity implements ItemRegistryEntity {
                 item = new Item(properties);
             }
             if (this.compostChance != null) ComposterBlock.add(this.compostChance, item);
+            KiterinoItemModifierImpl.MODIFIER_IMPL.add(new org.bukkit.NamespacedKey(this.itemKey.getNamespace(), this.itemKey.getPath()), buildModifier(item));
             return item;
         }
 
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjectorImpl.java b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjectorImpl.java
index a35b5265681c6c535fc5a72a548b4dbeb5d6afca..66be336c6c00b16c627f1b22586b590839f2ea4e 100644
--- a/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjectorImpl.java
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjectorImpl.java
@@ -140,6 +140,15 @@ public class KiterinoBootstrapMaterialInjectorImpl implements IKiterinoBootstrap
             Field enumConstantDirectory = Class.class.getDeclaredField("enumConstantDirectory");
             enumConstantDirectory.setAccessible(true);
             enumConstantDirectory.set(Material.class, null);
+
+            var modifier = me.sosedik.kiterino.modifier.item.KiterinoItemModifierImpl.MODIFIER_IMPL;
+            if (!me.sosedik.kiterino.KiterinoConfig.itemModifiersOrder.contains(modifier.getModifierId())) {
+                me.sosedik.kiterino.modifier.item.ItemModifiersHandlerImpl.modifiers.put(modifier.getModifierId(), modifier);
+                me.sosedik.kiterino.KiterinoConfig.itemModifiersOrder.addLast(modifier.getModifierId());
+                me.sosedik.kiterino.KiterinoConfig.log(java.util.logging.Level.WARNING, "Item modifier with id " + modifier.getModifierId() + " is missing from \"item-modifiers.modification-order\" option, but custom item was injected. Forcing the modifier automatically.");
+            } else {
+                modifier.register();
+            }
         } catch (Exception e) {
             throw new KiterinoMaterialExtenderRuntimeException("Couldn't extend Material", e);
         }
